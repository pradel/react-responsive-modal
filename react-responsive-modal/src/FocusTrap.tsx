import { useEffect, useRef } from 'react';
import { isBrowser } from './utils';
import {
  tabTrappingKey,
  candidateSelectors,
  getAllTabbingElements,
} from './lib/focusTrapJs';

interface FocusTrapProps {
  container?: React.RefObject<HTMLElement> | null;
}

export const FocusTrap = ({ container }: FocusTrapProps) => {
  const refLastFocus = useRef<HTMLElement | null>();
  /**
   * Handle focus lock on the modal
   */
  useEffect(() => {
    const handleKeyEvent = (event: KeyboardEvent) => {
      if (container?.current) {
        tabTrappingKey(event, container.current);
      }
    };

    if (isBrowser) {
      document.addEventListener('keydown', handleKeyEvent);
    }
    // On mount we focus on the first focusable element in the modal if there is one
    if (isBrowser && container?.current) {
      const allTabbingElements = getAllTabbingElements(container.current);
      if (allTabbingElements[0]) {
        // First we save the last focused element
        // only if it's a focusable element
        if (
          candidateSelectors.findIndex((selector) =>
            document.activeElement?.matches(selector)
          ) !== -1
        ) {
          refLastFocus.current = document.activeElement as HTMLElement;
        }
        allTabbingElements[0].focus();
      }
    }
    return () => {
      if (isBrowser) {
        document.removeEventListener('keydown', handleKeyEvent);
        // On unmount we restore the focus to the last focused element
        refLastFocus.current?.focus();
      }
    };
  }, [container]);

  return null;
};
