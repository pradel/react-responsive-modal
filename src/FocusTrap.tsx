import { useEffect } from 'react';
import { isBrowser } from './utils';
import * as focusTrapJs from './focusTrapJs';

interface FocusTrapProps {
  container?: React.RefObject<HTMLElement> | null;
}

export const FocusTrap = ({ container }: FocusTrapProps) => {
  /**
   * Handle focus lock on the modal
   */
  useEffect(() => {
    const handleKeyEvent = (event: KeyboardEvent) => {
      if (container?.current) {
        focusTrapJs.tabTrappingKey(event, container.current);
      }
    };

    if (isBrowser) {
      document.addEventListener('keydown', handleKeyEvent);
    }
    // On mount we focus on the first focusable element in the modal if there is one
    if (isBrowser && container?.current) {
      const allTabbingElements = focusTrapJs.getAllTabbingElements(
        container.current
      );
      if (allTabbingElements[0]) {
        allTabbingElements[0].focus();
      }
    }
    return () => {
      if (isBrowser) {
        document.removeEventListener('keydown', handleKeyEvent);
      }
    };
  }, [container]);

  return null;
};
