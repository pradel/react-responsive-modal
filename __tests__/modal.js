import React from 'react';
import { mount } from 'enzyme';
import FocusTrap from 'focus-trap-react';
import CSSTransition from 'react-transition-group/CSSTransition';
import Modal from '../src/modal';

const defaultProps = {
  classes: {
    overlay: 'test-react-responsive-modal-overlay',
    modal: 'test-react-responsive-modal-modal',
    modalCenter: 'test-react-responsive-modal-modal-center',
    closeButton: 'test-react-responsive-modal-close-button',
    closeIcon: 'test-react-responsive-modal-close-icon',
    transitionEnter: 'test-react-responsive-modal-transition-enter',
    transitionEnterActive:
      'test-react-responsive-modal-transition-enter-active',
    transitionExit: 'test-react-responsive-modal-transition-exit',
    transitionExitActive: 'test-react-responsive-modal-transition-exit-active',
  },
  open: false,
  onClose: jest.fn(),
  animationDuration: 50,
};

const mockEvent = {
  stopPropagation: jest.fn(),
  target: {},
};

const wait = (time = 100) => new Promise(resolve => setTimeout(resolve, time));

describe('modal', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockEvent.target = {};
  });

  it('should have a default export', () => {
    expect(typeof Modal).toBe('function');
  });

  describe('overlay', () => {
    it('should pass the animationDuration prop to react-transition-group', () => {
      const wrapper = mount(
        <Modal {...defaultProps} open animationDuration={123}>
          <div>modal content</div>
        </Modal>
      );

      const transitionWrapper = wrapper.find(CSSTransition);
      expect(transitionWrapper.length).toBe(1);
      expect(transitionWrapper.props().timeout).toBe(123);
      wrapper.unmount();
    });

    it('should call onClose when click on the overlay', () => {
      const wrapper = mount(
        <Modal {...defaultProps} open>
          <div>modal content</div>
        </Modal>
      );

      const overlayWrapper = wrapper.find(`.${defaultProps.classes.overlay}`);
      overlayWrapper.simulate('click');
      expect(defaultProps.onClose).toHaveBeenCalled();
      wrapper.unmount();
    });

    it('should disable the handler when closeOnOverlayClick is false', () => {
      const wrapper = mount(
        <Modal {...defaultProps} open closeOnOverlayClick={false}>
          <div>modal content</div>
        </Modal>
      );

      const overlayWrapper = wrapper.find(`.${defaultProps.classes.overlay}`);
      overlayWrapper.simulate('click');
      expect(defaultProps.onClose).not.toHaveBeenCalled();
      wrapper.unmount();
    });

    it('should ignore the overlay click if the event does not come from the overlay', () => {
      const wrapper = mount(
        <Modal {...defaultProps} open>
          <div>modal content</div>
        </Modal>
      );

      const modalWrapper = wrapper.find(`.${defaultProps.classes.modal}`);
      modalWrapper.simulate('click');
      expect(defaultProps.onClose).not.toHaveBeenCalled();
      wrapper.unmount();
    });
  });

  describe('esc key down', () => {
    it('an invalid event should not call onClose', () => {
      const wrapper = mount(
        <Modal {...defaultProps} open>
          <div>modal content</div>
        </Modal>
      );

      const handler = wrapper.instance().handleKeydown;
      handler({ keyCode: 10 });
      expect(defaultProps.onClose).not.toHaveBeenCalled();
      wrapper.unmount();
    });

    it('should not call onClose when closeOnEsc is false', () => {
      const wrapper = mount(
        <Modal {...defaultProps} open closeOnEsc={false}>
          <div>modal content</div>
        </Modal>
      );

      const handler = wrapper.instance().handleKeydown;
      handler({ keyCode: 27 });
      expect(defaultProps.onClose).not.toHaveBeenCalled();
      wrapper.unmount();
    });

    it('should call onClose', () => {
      const wrapper = mount(
        <Modal {...defaultProps} open>
          <div>modal content</div>
        </Modal>
      );

      const handler = wrapper.instance().handleKeydown;
      handler({ keyCode: 27 });
      expect(defaultProps.onClose).toHaveBeenCalled();
      wrapper.unmount();
    });
  });

  describe('closeIcon', () => {
    it('should hide closeIcon when showCloseIcon is false', () => {
      const wrapper = mount(
        <Modal
          {...defaultProps}
          open
          showCloseIcon={false}
          focusTrapped={false}
        >
          <div>modal content</div>
        </Modal>
      );

      expect(wrapper.find('CloseIcon').length).toBe(0);
      wrapper.unmount();
    });

    it('should call onClose', () => {
      const wrapper = mount(
        <Modal {...defaultProps} open>
          <div>modal content</div>
        </Modal>
      );

      wrapper.find('CloseIcon').simulate('click');
      expect(wrapper.find('CloseIcon').length).toBe(1);
      expect(defaultProps.onClose).toHaveBeenCalled();
      wrapper.unmount();
    });
  });

  describe('render', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = mount(
        <Modal {...defaultProps}>
          <div>modal content</div>
        </Modal>
      );
    });

    afterAll(() => {
      wrapper.unmount();
    });

    it('should render null when then modal is not opened', () => {
      expect(wrapper.find(Modal).length).toBe(1);
      expect(wrapper.find(Modal)).toBeEmptyRender();
    });

    it('should render the content', () => {
      wrapper.setProps({ open: true });
      expect(wrapper.find(Modal).length).toBe(1);
      expect(wrapper.find(Modal)).toMatchSnapshot();
    });
  });

  describe('lifecycle', () => {
    it('should show modal when prop open change to true', () => {
      const wrapper = mount(
        <Modal {...defaultProps}>
          <div>modal content</div>
        </Modal>
      );

      expect(wrapper.find(Modal)).toBeEmptyRender();
      expect(wrapper.state().showPortal).toBe(false);

      wrapper.setProps({ open: true });

      expect(wrapper.find(Modal).length).toBe(1);
      expect(wrapper.find(Modal)).not.toBeEmptyRender();
      expect(wrapper.state().showPortal).toBe(true);
      wrapper.unmount();
    });

    it('should hide modal when prop open change to false', async () => {
      const wrapper = mount(
        <Modal {...defaultProps} open>
          <div>modal content</div>
        </Modal>
      );

      expect(wrapper.find(Modal).length).toBe(1);
      expect(wrapper.find(Modal)).not.toBeEmptyRender();
      expect(wrapper.state().showPortal).toBe(true);

      wrapper.setProps({ open: false });
      await wait();

      expect(wrapper.find(Modal).html()).toBe(null);
      expect(wrapper.state().showPortal).toBe(false);
      wrapper.unmount();
    });
  });

  describe('block scroll', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = mount(
        <Modal {...defaultProps}>
          <div>modal content</div>
        </Modal>
      );
    });

    afterAll(() => {
      wrapper.unmount();
    });

    it('should not block the scroll when closed', () => {
      expect(document.documentElement.style.overflow).toBe('');
    });

    it('should block the scroll when opened', () => {
      wrapper.setProps({ open: true });
      expect(document.documentElement.style.overflow).toBe('hidden');
    });

    it('should unblock the scroll after animation', async () => {
      wrapper.setProps({ open: false });
      expect(document.documentElement.style.overflow).toBe('hidden');
      await wait();
      expect(document.documentElement.style.overflow).toBe('');
    });
  });

  describe('prop: onEntered', () => {
    it('should be called when component animation is finished', async () => {
      const onEntered = jest.fn();
      const wrapper = mount(
        <Modal {...defaultProps} open onEntered={onEntered}>
          <div>modal content</div>
        </Modal>
      );

      expect(onEntered).not.toHaveBeenCalled();
      await wait();

      expect(onEntered).toHaveBeenCalled();
      wrapper.unmount();
    });
  });

  describe('prop: onExited', () => {
    it('should be called when component animation is finished', async () => {
      const onExited = jest.fn();
      const wrapper = mount(
        <Modal {...defaultProps} open onExited={onExited}>
          <div>modal content</div>
        </Modal>
      );

      await wait();
      wrapper.setProps({ open: false });
      expect(onExited).not.toHaveBeenCalled();
      await wait();

      expect(onExited).toHaveBeenCalled();
      wrapper.unmount();
    });
  });

  describe('prop: onEscKeyDown', () => {
    it('should be called', async () => {
      const onEscKeyDown = jest.fn();
      const wrapper = mount(
        <Modal {...defaultProps} open onEscKeyDown={onEscKeyDown}>
          <div>modal content</div>
        </Modal>
      );

      const handler = wrapper.instance().handleKeydown;
      handler({ keyCode: 27 });
      expect(onEscKeyDown).toHaveBeenCalled();
      expect(defaultProps.onClose).toHaveBeenCalled();
      wrapper.unmount();
    });
  });

  describe('prop: onOverlayClick', () => {
    it('should be called', async () => {
      const onOverlayClick = jest.fn();
      const wrapper = mount(
        <Modal {...defaultProps} open onOverlayClick={onOverlayClick}>
          <div>modal content</div>
        </Modal>
      );

      const handler = wrapper.instance().handleClickOverlay;
      const overlayWrapper = wrapper.find(`.${defaultProps.classes.overlay}`);
      mockEvent.target.className = overlayWrapper.prop('className');
      handler(mockEvent);
      expect(onOverlayClick).toHaveBeenCalled();
      expect(defaultProps.onClose).toHaveBeenCalled();
      wrapper.unmount();
    });
  });

  describe('prop: center', () => {
    it('should apply center class to modal', async () => {
      const wrapper = mount(
        <Modal {...defaultProps} open center>
          <div>modal content</div>
        </Modal>
      );

      const modalWrapper = wrapper.find(`.${defaultProps.classes.modal}`);
      expect(
        modalWrapper.hasClass(defaultProps.classes.modalCenter)
      ).toBeTruthy();
      wrapper.unmount();
    });
  });

  describe('prop: classes', () => {
    it('should apply custom classes to the modal', async () => {
      const wrapper = mount(
        <Modal {...defaultProps} open center>
          <div>modal content</div>
        </Modal>
      );

      expect(wrapper.find(`.${defaultProps.classes.overlay}`)).toBeTruthy();
      expect(wrapper.find(`.${defaultProps.classes.modal}`)).toBeTruthy();
      expect(wrapper.find(`.${defaultProps.classes.modalCenter}`)).toBeTruthy();
      expect(wrapper.find(`.${defaultProps.classes.closeButton}`)).toBeTruthy();
      expect(wrapper.find(`.${defaultProps.classes.closeIcon}`)).toBeTruthy();
      wrapper.unmount();
    });
  });

  describe('prop: blockScroll', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = mount(
        <Modal {...defaultProps} blockScroll={false}>
          <div>modal content</div>
        </Modal>
      );
    });

    afterAll(() => {
      wrapper.unmount();
    });

    it('should not block the scroll when closed', () => {
      expect(document.documentElement.style.overflow).toBe('');
    });

    it('should not block the scroll when opened', () => {
      wrapper.setProps({ open: true });
      expect(document.documentElement.style.overflow).toBe('');
    });

    it('should not block the scroll before or after animation', async () => {
      wrapper.setProps({ open: false });
      expect(document.documentElement.style.overflow).toBe('');
      await wait();
      expect(document.documentElement.style.overflow).toBe('');
    });
  });

  describe('prop: focusTrapped', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = mount(
        <Modal {...defaultProps} open>
          <div>modal content</div>
        </Modal>
      );
    });

    afterAll(() => {
      wrapper.unmount();
    });

    it('should contain focus trap component when true (default)', () => {
      expect(wrapper.find(FocusTrap).exists()).toBe(true);
    });

    it('should not contain focus trap component when false', () => {
      wrapper.setProps({ focusTrapped: false });
      expect(wrapper.find(FocusTrap).exists()).toBe(false);
    });

    describe('prop: focusTrapOptions', () => {
      it('should have a default option for clickOutsideDeactivates true', () => {
        wrapper.setProps({ focusTrapped: true });
        expect(wrapper.find(FocusTrap).prop('focusTrapOptions')).toEqual({
          clickOutsideDeactivates: true,
        });
      });

      it('should override clickOutsideDeactivates if specified', () => {
        const focusTrapOptions = {
          clickOutsideDeactivates: false,
        };
        wrapper.setProps({ focusTrapOptions, focusTrapped: true });
        expect(wrapper.find(FocusTrap).prop('focusTrapOptions')).toEqual(
          focusTrapOptions
        );
      });

      it('should pass focusTrapOptions', () => {
        const focusTrapOptions = {
          clickOutsideDeactivates: false,
          escapeDeactivates: false,
          returnFocusOnDeactivate: false,
        };
        wrapper.setProps({ focusTrapOptions, focusTrapped: true });
        expect(wrapper.find(FocusTrap).prop('focusTrapOptions')).toEqual(
          focusTrapOptions
        );
      });
    });
  });
});
