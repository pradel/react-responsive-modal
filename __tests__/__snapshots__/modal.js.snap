// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modal render should render the content 1`] = `
<Modal
  animationDuration={50}
  blockScroll={true}
  center={false}
  classNames={Object {}}
  classes={
    Object {
      "closeButton": "test-react-responsive-modal-close-button",
      "closeIcon": "test-react-responsive-modal-close-icon",
      "modal": "test-react-responsive-modal-modal",
      "overlay": "test-react-responsive-modal-overlay",
      "overlayCenter": "test-react-responsive-modal-overlay-center",
      "transitionEnter": "test-react-responsive-modal-transition-enter",
      "transitionEnterActive": "test-react-responsive-modal-transition-enter-active",
      "transitionExit": "test-react-responsive-modal-transition-exit",
      "transitionExitActive": "test-react-responsive-modal-transition-exit-active",
    }
  }
  closeIconSize={28}
  closeIconSvgPath={
    <path
      d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
    />
  }
  closeOnEsc={true}
  closeOnOverlayClick={true}
  focusTrapOptions={Object {}}
  focusTrapped={true}
  onClose={[MockFunction]}
  open={true}
  role="dialog"
  showCloseIcon={true}
  styles={Object {}}
>
  <Portal
    containerInfo={
      <div>
        <div
          class="test-react-responsive-modal-overlay test-react-responsive-modal-transition-enter test-react-responsive-modal-transition-enter-active"
        >
          <div
            aria-modal="true"
            class="test-react-responsive-modal-modal"
            role="dialog"
          >
            <div>
              <div>
                modal content
              </div>
              <button
                class="test-react-responsive-modal-close-button"
              >
                <svg
                  class="test-react-responsive-modal-close-icon"
                  height="28"
                  viewBox="0 0 36 36"
                  width="28"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    }
  >
    <CSSTransition
      appear={true}
      classNames={
        Object {
          "appear": "test-react-responsive-modal-transition-enter",
          "appearActive": "test-react-responsive-modal-transition-enter-active",
          "enter": "test-react-responsive-modal-transition-enter",
          "enterActive": "test-react-responsive-modal-transition-enter-active",
          "exit": "test-react-responsive-modal-transition-exit",
          "exitActive": "test-react-responsive-modal-transition-exit-active",
        }
      }
      in={true}
      onEntered={[Function]}
      onExited={[Function]}
      timeout={50}
    >
      <Transition
        appear={true}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={50}
        unmountOnExit={false}
      >
        <div
          className="test-react-responsive-modal-overlay"
          onClick={[Function]}
        >
          <div
            aria-modal="true"
            className="test-react-responsive-modal-modal"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            role="dialog"
          >
            <FocusTrap
              _createFocusTrap={[Function]}
              active={true}
              focusTrapOptions={
                Object {
                  "clickOutsideDeactivates": true,
                }
              }
              paused={false}
              tag="div"
            >
              <div>
                <div>
                  modal content
                </div>
                <CloseIcon
                  classNames={Object {}}
                  classes={
                    Object {
                      "closeButton": "test-react-responsive-modal-close-button",
                      "closeIcon": "test-react-responsive-modal-close-icon",
                      "modal": "test-react-responsive-modal-modal",
                      "overlay": "test-react-responsive-modal-overlay",
                      "overlayCenter": "test-react-responsive-modal-overlay-center",
                      "transitionEnter": "test-react-responsive-modal-transition-enter",
                      "transitionEnterActive": "test-react-responsive-modal-transition-enter-active",
                      "transitionExit": "test-react-responsive-modal-transition-exit",
                      "transitionExitActive": "test-react-responsive-modal-transition-exit-active",
                    }
                  }
                  closeIconSize={28}
                  closeIconSvgPath={
                    <path
                      d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
                    />
                  }
                  id={null}
                  onClickCloseIcon={[Function]}
                  styles={Object {}}
                >
                  <button
                    className="test-react-responsive-modal-close-button"
                    id={null}
                    onClick={[Function]}
                  >
                    <svg
                      className="test-react-responsive-modal-close-icon"
                      height={28}
                      viewBox="0 0 36 36"
                      width={28}
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
                      />
                    </svg>
                  </button>
                </CloseIcon>
              </div>
            </FocusTrap>
          </div>
        </div>
      </Transition>
    </CSSTransition>
  </Portal>
</Modal>
`;
